Event 기능 구현

yuyoungrhee opened the pull request:
### 주요 변경 사항
EventService 클래스 추가:
- 이벤트 목록 조회, 이벤트 상세 조회 및 좌석 조회 기능 구현
- DTO 변환 로직을 별도로 분리하여 코드 중복 최소화

EventController 클래스 추가:
- 이벤트 목록 조회, 상세 정보 조회 및 좌석 조회 엔드포인트 구현
- 이벤트 ID 유효성 검사로 예외 처리

Seat 엔티티 수정 사항
- Order의 null을 허용하도록  nullable = false 삭제: 좌석이 예약되지 않은 초기 상태에서는 order가 null상태임

### 구현한 API
- GET /events: 모든 이벤트 목록 조회
- GET /events/{eventId}: 특정 이벤트 상세 조회
- GET /events/{eventId}/seats: 특정 이벤트의 좌석 정보 조회

### 테스트
- 단위 테스트 및 통합 테스트 추가: EventService 및 EventController의 메소드 검증

### 궁금한 점
- Event 엔티티에 String 타입의 duration 컬럼이 있는데, 이거는 공연 기간 날짜를 말하는 걸까요? 이대로 둬도 되는지, 아니면 시작 날짜, 종료 날짜 컬럼을 나눠서 추가해야 할지.
- Event 엔티티에 ticketCloseTime도 있어야 하지 않나요?

yuyoungrhee committed:
feat: 공연 목록 전체 조회 API 구현

- 이벤트 컨트롤러, 서비스, 레파지토리 작성
- JPA 기반 EventRepository 인터페이스로 CRUD 지원
- ResponseDto 클래스 추가

yuyoungrhee committed:
test: EventSerivce의 getAllEvents() 테스트 코드 작성

- EventService의 전체 이벤트 조회 기능에 대한 테스트 추가
- 여러 이벤트 저장 후 조회 테스트 구현
- 테스트 데이터 생성 메서드(createEvent) 추가

yuyoungrhee committed:
feat: Event 상세 조회 API 구현

- EventController에 공연 상세 조회 API 엔드포인트 추가
- EventService에서 특정 이벤트 ID로 공연 상세 정보 조회하는 로직 구현
- EventResponseDto를 통해 필요한 공연 정보를 반환

yuyoungrhee committed:
feat: 특정 Event의 좌석 조회 API 구현

- EventController에 좌석 전체 조회 API 엔드포인트 추가
- EventService에서 특정 이벤트 ID로 좌석 전체를 조회하는 로직 구현
- 데이터베이스 접근을 위한 JPA기반 SeatRepository 생성
- 좌석 데이터를 전송하기 위한 SeatResponseDto 생성

yuyoungrhee committed:
refactor: EventResponseDto의 생성자 접근 제한

- EventResponseDto 클래스의 기본 생성자를 PROTECTED 접근으로 변경하여 외부에서 인스턴스를 직접 new로 생성할 수 없도록 수정
- Builder 패턴을 통해서만 인스턴스 생성이 가능하도록 개선

yuyoungrhee committed:
test: EventService 및 EventController 테스트 코드 작성

- EventService의 메소드에 대한 단위 테스트 추가
- EventController의 API 엔드포인트에 대한 통합 테스트 추가

yuyoungrhee committed:
refactor: EventService에서 DTO 변환 로직 분리 및 Optional 반환

- 엔티티를 Response DTO로 변환하는 로직을 별도의 메서드로 분리하여 코드 중복을 줄임.
- getEventById() 메서드를 DTO를 직접 반환하는 대신 Optional<EventResponseDto>를 반환하도록 수정.
- Service 수정 내용에 맞춰 ServiceTest 코드도 수정

yuyoungrhee committed:
refactor: EventController 예외 처리 추가 및 엔드포인트 수정

- getEventById() 메서드에서 이벤트 ID가 들어오지 않을 경우 예외를 반환하도록 처리 추가
- 상세 정보 조회 엔드포인트를 /{eventId}에서 /events/{eventId}로 일관되게 수정

asdf56782 commented on src/main/java/com/example/goorm_ticket/api/event/service/EventService.java:
```diff
@@ -68,4 +61,11 @@ public List<SeatResponseDto> getSeatsByEventId(Long eventId) {
     }
 
+    private EventResponseDto mapToEventResponseDto(Event event) {
+        return EventResponseDto.builder()
```
### DTO 변환 로직의 위치

Event 객체를 기반으로 EventResponseDto 객체를 생성하는 부분은 EventResponseDto 도메인 내부에 작성하는 것을 추천합니다!

yuyoungrhee replied:
수정했습니다!

wjd4204 commented on src/main/java/com/example/goorm_ticket/api/event/controller/EventController.java:
```diff
@@ -0,0 +1,51 @@
+@RestController
+public class EventController {
+
+    private final EventService eventService;
```
어노테이션 RequiredArgsConstructor를 사용하여 생성자 주입을 편리하게 해주는 것이 좋습니다!

yuyoungrhee replied:
`@RequiredArgsConstructor` 추가하고 기존 `@Autowired`한 생성자를 없애는 쪽으로 수정했습니다!

wjd4204 replied:
메서드마다 api를 정하기 보다는 클래스 위에 RequestMapping을 써서 공통 부분은 한 번에 처리하는 것이 좋아보입니다!

wjd4204 commented on src/main/java/com/example/goorm_ticket/api/event/controller/EventController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    // 전체 페이지 이벤트 조회 API
+    @GetMapping("/events")
```
api를 명시할 때는 보통 '/api/{버전명}/기능명' 으로 하는 것이 좋습니다!

yuyoungrhee replied:
`@GetMapping("/api/v1/events")` 로 수정했습니다. 나머지 API도 같이 수정했습니다.

wjd4204 commented on src/main/java/com/example/goorm_ticket/api/event/controller/EventController.java:
```diff
@@ -0,0 +1,51 @@
+    // 이벤트ID로 상세 정보 조회
+    @GetMapping("/events/{eventId}")
+    public ResponseEntity<EventResponseDto> getEventById(@PathVariable Long eventId) {
+        EventResponseDto event = eventService.getEventById(eventId)
```
controller에서 exception을 발행하는 행위는 controller 클래스 특징에 어울리지 않습니다. 서비스 클래스에서 이를 확인하여 반환하는 것이 좋아 보입니다!

asdf56782 replied:
@ExceptionHandler 사용하면 서비스단에서 발생한 예외에 대한 응답을 생성할 수 있습니다!

yuyoungrhee replied:
@ExceptionHandler를 사용해서 `GlobalExceptionHandler` 클래스를 만들어 예외처리를 하는 방법으로 수정했습니다.
머지할 때 주의가 필요해 보입니다..

yuyoungrhee committed:
refactor: 컨트롤러 예외 처리 로직을 서비스 계층으로 이동 및 테스트 코드 수정

- 컨트롤러에서 예외를 처리하던 로직을 제거하고, 서비스 계층에서 처리하도록 변경
  - 컨트롤러가 비즈니스 로직을 직접 다루지 않도록 하고 서비스가 모든 예외 처리를 담당
  - EventService에서 `Optional`을 반환하지 않고, 예외를 직접 던지는 방식으로 수정
- 테스트 코드 수정
  - 기존 `Optional`을 사용하던 테스트 코드를, 예외 발생 여부를 검증하도록 수정
  - 성공적인 이벤트 조회 테스트 코드에서 `Optional` 제거 후 `EventResponseDto` 바로 검증

yuyoungrhee committed:
refactor: @RequiredArgsConstructor를 사용하여 컨트롤러에 의존성 주입

- 기존 @Autowired를 통해 명시적으로 생성하던 생성자를 제거하고, final 필드를 사용한 간결한 생성자 주입으로 변경

yuyoungrhee committed:
refactor: API 엔드포인트를 '/api/v1/{버전명}/기능명' 형식으로 변경

yuyoungrhee committed:
refactor: Dto 내부에 Dto 변환 로직을 두도록 위치 수정

- 기존 EventService에 정의되어있던 DTO 변환 로직(mapToEventResponseDto)을 EventResponseDto 클래스 내부로 이동.
- 변환 메서드를 static으로 정의하여 인스턴스 생성 없이 호출 가능하도록 함.

yuyoungrhee committed:
feat: GlobalExceptionHandler를 통한 컨트롤러 전역의 예외 처리 추가

- 모든 컨트롤러에서 공통 예외를 처리하기 위해 GlobalExceptionHandler 클래스를 추가
- NoSuchElementException 및 IllegalArgumentException에 대한 구체적인 예외 처리를 구현

wjd4204 commented on src/main/java/com/example/goorm_ticket/api/event/service/EventService.java:
```diff
@@ -0,0 +1,63 @@
+
+
+    // 페이징된 이벤트 리스트 조회, DTO 변환
+    public Page<EventResponseDto> getAllEvents(int page, int size) {
```
데이터를 수정하지 않는다면 Transaction(readOnly = true)로 해도 되지 않나여?

yuyoungrhee replied:
그러네요 
현재 EventService에는 조회 메서드만 있기 때문에 Service클래스 상단에 @Transactional(readOnly = true)를 전체적으로 추가했습니다.

wjd4204 commented on src/main/java/com/example/goorm_ticket/domain/event/repository/EventRepository.java:
```diff
@@ -0,0 +1,7 @@
+
+import com.example.goorm_ticket.domain.event.entity.Event;
+import org.springframework.data.jpa.repository.JpaRepository;
+
```
어노테이션 Repository를 붙히는 것이 좋아보입니다!

yuyoungrhee replied:
어머 왜 안붙였는지 모르겠네요.. 추가했습니다

wjd4204 commented on src/test/java/com/example/goorm_ticket/api/event/controller/EventControllerTest.java:
```diff
@@ -0,0 +1,135 @@
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@WebMvcTest(EventController.class)
```
WebMvcTest와 SpringBootTest와의 차이점이 무엇인가요??

yuyoungrhee replied:
@WebMvcTest(EventController.class) 는 EventController.class에 필요한 빈만 로드함기 때문에 상대적으로 속도가 빠릅니다.

현재 코드는 **컨트롤러 테스트**에 초점을 맞춰서 요청 및 응답을 테스트하고 있습니다.
실제 컨트롤러와 서비스 간의 상호작용을 포함하고 있지만 컨트롤러의 동작을 독립적으로 검증하는 단위테스트와 비슷한 성격입니다.

따라서 속도가 더 빠른 WebMvcTest를 사용했고, 테스트코드 강의에서도 컨트롤러테스트는 WebMvcTest 환경으로 통일했습니다.

yuyoungrhee committed:
feat: 레파지토리에 @Repository 어노테이션 추가

- 스프링 빈으로 등록하기 위하여 @Repository 어노테이션 추가

yuyoungrhee committed:
feat: EventService 클래스에 @Transactional(readOnly = true) 추가

- 클래스 상단에 @Transactional(readOnly = true) 추가
- 기존에 getAllEvents 메서드에만 readOnly 설정이 없었고, 현재 EventService에는 조회 메서드만 존재하기때문에 상단으로 옮겼습니다.
